File: appium-mobile-tests/reporters/excelReporter.js
/**
 * Excel Report Generator for Appium Mobile E2E Test Suite
 * Generates formatted .xlsx Excel analysis workbook with summary statistics,
 * detailed test execution logs, and screen coverage metrics.
 */

const ExcelJS = require('exceljs');
const fs = require('fs');
const path = require('path');

class ExcelReporter {
  constructor(reportDir = path.join(__dirname, '..', 'reports')) {
    this.reportDir = reportDir;
    if (!fs.existsSync(this.reportDir)) {
      fs.mkdirSync(this.reportDir, { recursive: true });
    }
  }

  async generateReport(testResults, deviceInfo = {}) {
    const workbook = new ExcelJS.Workbook();
    workbook.creator = 'FuelGo Appium Automation Engine';
    workbook.created = new Date();

    // ----------------------------------------------------
    // TAB 1: EXECUTIVE SUMMARY
    // ----------------------------------------------------
    const summarySheet = workbook.addWorksheet('Dashboard & Summary', {
      views: [{ showGridLines: true }]
    });

    // Title Banner
    summarySheet.mergeCells('A1:E2');
    const titleCell = summarySheet.getCell('A1');
    titleCell.value = '📱 FuelGo Android Appium E2E Automation Report';
    titleCell.font = { name: 'Arial', size: 16, bold: true, color: { argb: 'FFFFFF' } };
    titleCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: '0B1A3B' } }; // FuelGo Navy
    titleCell.alignment = { vertical: 'middle', horizontal: 'center' };

    // Metadata Table
    const totalTests = testResults.length;
    const passedTests = testResults.filter(t => t.status === 'PASS').length;
    const failedTests = testResults.filter(t => t.status === 'FAIL').length;
    const passRate = totalTests > 0 ? ((passedTests / totalTests) * 100).toFixed(1) : 0;
    const totalDuration = (testResults.reduce((acc, t) => acc + (t.duration || 0), 0) / 1000).toFixed(2);

    summarySheet.addRow([]); // Blank row
    summarySheet.addRow(['Execution Date', new Date().toLocaleString()]);
    summarySheet.addRow(['Target Device', deviceInfo.deviceName || 'Android Emulator (UiAutomator2)']);
    summarySheet.addRow(['OS Platform', `${deviceInfo.platformName || 'Android'} ${deviceInfo.platformVersion || '13.0'}`]);
    summarySheet.addRow(['Application URL / APK', deviceInfo.app || 'FuelGo Mobile Webview/Native']);
    summarySheet.addRow([]);

    // KPI Metrics Card Table
    summarySheet.addRow(['Metric', 'Value', 'Status Benchmark']);
    const kpiHeader = summarySheet.getRow(9);
    kpiHeader.font = { bold: true, color: { argb: 'FFFFFF' } };
    kpiHeader.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: '1565D8' } };

    const metricsData = [
      ['Total Test Cases', totalTests, 'Target: 100% Executed'],
      ['Passed Test Cases', passedTests, 'Target: All Pass'],
      ['Failed Test Cases', failedTests, failedTests === 0 ? 'Optimal (0)' : 'Requires Review'],
      ['Success Pass Rate', `${passRate}%`, parseFloat(passRate) >= 90 ? 'PASSED HIGH' : 'CRITICAL ATTENTION'],
      ['Total Duration (sec)', `${totalDuration}s`, 'Performance Nominal']
    ];

    metricsData.forEach(row => {
      const addedRow = summarySheet.addRow(row);
      if (row[0] === 'Passed Test Cases') {
        addedRow.getCell(2).font = { color: { argb: '008000' }, bold: true };
      } else if (row[0] === 'Failed Test Cases' && failedTests > 0) {
        addedRow.getCell(2).font = { color: { argb: 'FF0000' }, bold: true };
      }
    });

    summarySheet.getColumn('A').width = 25;
    summarySheet.getColumn('B').width = 35;
    summarySheet.getColumn('C').width = 30;

    // ----------------------------------------------------
    // TAB 2: DETAILED TEST RESULTS
    // ----------------------------------------------------
    const detailsSheet = workbook.addWorksheet('Test Execution Details', {
      views: [{ showGridLines: true }]
    });

    detailsSheet.columns = [
      { header: '#', key: 'id', width: 6 },
      { header: 'Test Suite / Module', key: 'suite', width: 22 },
      { header: 'Test Case Description', key: 'title', width: 45 },
      { header: 'Target Screen', key: 'screen', width: 18 },
      { header: 'Duration (ms)', key: 'duration', width: 15 },
      { header: 'Status', key: 'status', width: 12 },
      { header: 'Error / Failure Details', key: 'error', width: 45 }
    ];

    // Header Styling
    const headerRow = detailsSheet.getRow(1);
    headerRow.font = { name: 'Arial', size: 11, bold: true, color: { argb: 'FFFFFF' } };
    headerRow.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF6B00' } }; // FuelGo Orange
    headerRow.alignment = { vertical: 'middle', horizontal: 'center' };

    // Fill Test Data
    testResults.forEach((t, idx) => {
      const row = detailsSheet.addRow({
        id: idx + 1,
        suite: t.suite || 'E2E Flow',
        title: t.title,
        screen: t.screen || 'General',
        duration: t.duration || 0,
        status: t.status,
        error: t.error || 'N/A'
      });

      const statusCell = row.getCell('status');
      if (t.status === 'PASS') {
        statusCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'E8F5E9' } };
        statusCell.font = { color: { argb: '2E7D32' }, bold: true };
      } else {
        statusCell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFEBEE' } };
        statusCell.font = { color: { argb: 'C62828' }, bold: true };
      }
    });

    // ----------------------------------------------------
    // TAB 3: SCREEN COVERAGE ANALYSIS
    // ----------------------------------------------------
    const coverageSheet = workbook.addWorksheet('Screen Coverage Analysis');
    coverageSheet.columns = [
      { header: 'Screen Name', key: 'screen', width: 25 },
      { header: 'Total Checks', key: 'total', width: 15 },
      { header: 'Passed', key: 'passed', width: 15 },
      { header: 'Failed', key: 'failed', width: 15 },
      { header: 'Health Status', key: 'health', width: 20 }
    ];

    const covHeader = coverageSheet.getRow(1);
    covHeader.font = { bold: true, color: { argb: 'FFFFFF' } };
    covHeader.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: '0B1A3B' } };

    const screenMap = {};
    testResults.forEach(t => {
      const scr = t.screen || 'General';
      if (!screenMap[scr]) screenMap[scr] = { total: 0, passed: 0, failed: 0 };
      screenMap[scr].total++;
      if (t.status === 'PASS') screenMap[scr].passed++;
      else screenMap[scr].failed++;
    });

    Object.keys(screenMap).forEach(scr => {
      const item = screenMap[scr];
      const health = item.failed === 0 ? 'HEALTHY (100%)' : `${((item.passed / item.total) * 100).toFixed(0)}% PASSED`;
      coverageSheet.addRow({
        screen: scr,
        total: item.total,
        passed: item.passed,
        failed: item.failed,
        health: health
      });
    });

    // Save Workbook to File
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const fileName = `Appium_E2E_Test_Report_${timestamp}.xlsx`;
    const filePath = path.join(this.reportDir, fileName);

    await workbook.xlsx.writeFile(filePath);
    console.log(`\n✅ Excel Analysis Report generated successfully:`);
    console.log(`📂 Path: ${filePath}`);

    return filePath;
  }
}

module.exports = ExcelReporter;

File: appium-mobile-tests/python/excel_reporter.py
import os
from datetime import datetime
from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter

class PythonExcelReporter:
    def __init__(self, report_dir=None):
        if report_dir is None:
            report_dir = os.path.join(os.path.dirname(__file__), "..", "reports")
        self.report_dir = report_dir
        os.makedirs(self.report_dir, exist_ok=True)

    def generate_report(self, test_results, device_info=None):
        if device_info is None:
            device_info = {
                "deviceName": "Android Emulator (UiAutomator2)",
                "platformName": "Android",
                "platformVersion": "13.0",
                "app": "FuelGo Mobile App"
            }

        wb = Workbook()
        
        # ----------------------------------------------------
        # TAB 1: SUMMARY DASHBOARD
        # ----------------------------------------------------
        ws_summary = wb.active
        ws_summary.title = "Executive Summary"
        ws_summary.views.sheetView[0].showGridLines = True

        # Header Title
        ws_summary.merge_cells('A1:E2')
        title_cell = ws_summary['A1']
        title_cell.value = "📱 FuelGo Android Appium E2E Automation Report (Python)"
        title_cell.font = Font(name="Arial", size=14, bold=True, color="FFFFFF")
        title_cell.fill = PatternFill(start_color="0B1A3B", end_color="0B1A3B", fill_type="solid")
        title_cell.alignment = Alignment(horizontal="center", vertical="center")

        total_tests = len(test_results)
        passed_tests = sum(1 for t in test_results if t.get('status') == 'PASS')
        failed_tests = sum(1 for t in test_results if t.get('status') == 'FAIL')
        pass_rate = round((passed_tests / total_tests) * 100, 1) if total_tests > 0 else 0
        total_duration = round(sum(t.get('duration', 0) for t in test_results) / 1000, 2)

        ws_summary.append([])
        ws_summary.append(["Execution Date", datetime.now().strftime("%Y-%m-%d %H:%M:%S")])
        ws_summary.append(["Target Device", device_info.get("deviceName")])
        ws_summary.append(["Platform", f"{device_info.get('platformName')} {device_info.get('platformVersion')}"])
        ws_summary.append(["App / Package", device_info.get("app")])
        ws_summary.append([])

        ws_summary.append(["KPI Metric", "Result Value", "Benchmark Target"])
        kpi_header_row = ws_summary[9]
        for cell in kpi_header_row:
            cell.font = Font(bold=True, color="FFFFFF")
            cell.fill = PatternFill(start_color="1565D8", end_color="1565D8", fill_type="solid")

        kpis = [
            ["Total Test Cases", total_tests, "Target: 100% Executed"],
            ["Passed Test Cases", passed_tests, "Target: All Pass"],
            ["Failed Test Cases", failed_tests, "0 Failures"],
            ["Pass Rate (%)", f"{pass_rate}%", "Threshold: >= 90%"],
            ["Total Duration (sec)", f"{total_duration}s", "Performance Nominal"]
        ]

        for row in kpis:
            ws_summary.append(row)

        ws_summary.column_dimensions['A'].width = 25
        ws_summary.column_dimensions['B'].width = 30
        ws_summary.column_dimensions['C'].width = 25

        # ----------------------------------------------------
        # TAB 2: TEST DETAILS
        # ----------------------------------------------------
        ws_details = wb.create_sheet(title="Detailed Test Logs")
        ws_details.views.sheetView[0].showGridLines = True

        headers = ["#", "Suite", "Test Case Description", "Screen", "Duration (ms)", "Status", "Error Details"]
        ws_details.append(headers)

        header_row = ws_details[1]
        for cell in header_row:
            cell.font = Font(bold=True, color="FFFFFF")
            cell.fill = PatternFill(start_color="FF6B00", end_color="FF6B00", fill_type="solid")
            cell.alignment = Alignment(horizontal="center")

        for idx, t in enumerate(test_results, start=1):
            row_data = [
                idx,
                t.get('suite', 'E2E Suite'),
                t.get('title'),
                t.get('screen', 'General'),
                t.get('duration', 0),
                t.get('status'),
                t.get('error', 'N/A')
            ]
            ws_details.append(row_data)
            
            # Color status cell
            status_cell = ws_details.cell(row=idx+1, column=6)
            if t.get('status') == 'PASS':
                status_cell.fill = PatternFill(start_color="E8F5E9", end_color="E8F5E9", fill_type="solid")
                status_cell.font = Font(color="2E7D32", bold=True)
            else:
                status_cell.fill = PatternFill(start_color="FFEBEE", end_color="FFEBEE", fill_type="solid")
                status_cell.font = Font(color="C62828", bold=True)

        col_widths = {'A': 6, 'B': 20, 'C': 42, 'D': 18, 'E': 15, 'F': 12, 'G': 40}
        for col, width in col_widths.items():
            ws_details.column_dimensions[col].width = width

        timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
        file_name = f"Python_Appium_E2E_Report_{timestamp}.xlsx"
        file_path = os.path.join(self.report_dir, file_name)

        wb.save(file_path)
        print(f"\n✅ Python Excel Analysis Report generated successfully:")
        print(f"📂 Path: {file_path}")
        return file_path
